misc/pyudmi/src/udmi/client/manager/discovery/discovery_manager.py

```diff
@@ -0,0 +1,382 @@
+
+        # Set the duration to match the interval duration so that the scheduled
+        # stop logic is simpler
+        # if neither are set, this sets it to 0 (the default)
+        scan_duration_sec = config.scan_duration_sec if \
+            config.scan_duration_sec is not None else scan_interval_sec
+
+        while not self.scheduler_thread_stopped.is_set():
+            if time.monotonic() > next_action_time:
+                with self.mutex:
+                    if config != self.config:
+                        # Check that the config has not changed whilst the
+                        # scheduler was waiting to acquire the lock
+                        logging.info("config has change, exiting")
+                        return
+
+                    # make a copy of the current action so next_action can be
+                    # safely mutated
+                    current_action = next_action
+
+                    if current_action == ACTION_START:
+                        self._start()
+
+                        if scan_duration_sec > 0:
+                            next_action = ACTION_STOP
+                            next_action_time = (time.monotonic() +
+                                                scan_duration_sec)
+                            logging.info(
+                                f"scheduled discovery stop for "
+                                f"{self.__class__.__name__} in "
+                                f"{scan_duration_sec} seconds")
+                        else:
+                            # the scan runs indefinitely, exit the scheduler
+                            logging.info(
+                                f"{self.__class__.__name__} discovery running "
+                                f"indefinitely")
+                            return
+
+                    elif current_action == ACTION_STOP:
+                        self._stop()
+
+                        # If the scan is repetitive, schedule the next start
+                        if scan_interval_sec > 0:
+                            next_action = ACTION_START
+                            sleep_interval = (scan_interval_sec -
+                                              scan_duration_sec)
+                            next_action_time = time.monotonic() + sleep_interval
+                            logging.info(
+                                f"scheduled discovery start for "
+                                f"{self.__class__.__name__} in "
+                                f"{scan_duration_sec} seconds")
+                            self._set_internal_status(Status.SCHEDULED)
+                            self.state.phase = Phase.pending
+                        else:
+                            # The scan is not repetitive, exit the scheduler
+                            return
+
+            time.sleep(RUNNER_LOOP_INTERVAL)
+
+    @catch_exceptions_to_status
+    def apply_discovery_config(self, config_dict):
```

Comment:
is there a place in here where there's an example of how this dict is created?  I can guess (likely just a standard json-to-dict parsing), but I wanted to get it codified somewhere to be 100% sure. I saw the `self.publish(event.to_json())` code elsewhere, so I'm looking for the inverse...

(also, as an abstraction thing I'd expect the input/output paths to be roughly similar, so that the concept of "json" is either entirely in this layer, or not... so if JSON is output here then it would also be parsed here. I.e. it produces the same kind of thing that it consumes!)